The computed value of float with absolute positioning when there is no box should be "none"

Ahmad-S792 opened the pull request:
#### 6c209adee5ff65d632a60c344c5b4555bfbfb5df
<pre>
The computed value of float with absolute positioning when there is no box should be &quot;none&quot;

<a href="https://bugs.webkit.org/show_bug.cgi?id=286882">https://bugs.webkit.org/show_bug.cgi?id=286882</a>
<a href="https://rdar.apple.com/144045558">rdar://144045558</a>

Reviewed by Alan Baradlay.

This patch aligns WebKit with Web Specification [1]:

[1] <a href="https://www.w3.org/TR/CSS2/visuren.html#dis-pos-flo">https://www.w3.org/TR/CSS2/visuren.html#dis-pos-flo</a>

In this patch, we return computed value of `float: none` irrespective of
whether the display is `none` or not. Which is aligned with specification,
which mentions:

&quot;If &apos;display&apos; has the value &apos;none&apos;, then &apos;position&apos; and &apos;float&apos; do not apply.&quot;

and similarly if position is absolute or fixed (hasOutOfFlowPosition) function,
then it should also be float value should be `none`.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(ComputedStyleExtractor::valueForPropertyInStyle):
* LayoutTests/imported/w3c/web-platform-tests/css/CSS2/floats/computed-float-position-absolute-expected.txt:

Canonical link: <a href="https://commits.webkit.org/290061@main">https://commits.webkit.org/290061@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/84119e0a4632e20094972d6c640321ad81e7671d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88871 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/8395 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/55/builds/43339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/93841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/39630 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90922 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/8782 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/123/builds/16579 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/93841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/39630 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91873 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/8782 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/55/builds/43339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/93841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/8782 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/55/builds/43339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/38738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/8782 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/43339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/95681 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/16051 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/123/builds/16579 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/95681 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/16307 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/55/builds/43339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/95681 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/43339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/9117 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13910 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/16065 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/21376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/15806 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/19257 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/17587 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/style/StyleAdjuster.cpp:
```diff
@@ -421,6 +421,9 @@ void Adjuster::adjust(RenderStyle& style, const RenderStyle* userAgentAppearance
         style.setEffectiveDisplay(DisplayType::Block);
     }
 
+    if (style.display() == DisplayType::None && style.position() == PositionType::Absolute)
```
I wonder if we should do this check at `getComputedStyle` time instead, so we don't pay the cost of running this code on every single style adjust.

Ahmad-S792 replied:
@smfr - Done and locally tested that we are progressing WPT still with new change. Thanks for your input and help!

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/CSS2/floats/computed-float-position-absolute.html:
Please make sure import-expectations.json reflects that this is imported.

Ahmad-S792 replied:
Will do.

nt1m reviewed (commented):
What is the part of the spec that mandates this? position: absolute + display: none seems oddly specific.

Commit:
The computed value of float with absolute positioning when there is no box should be "none"

https://bugs.webkit.org/show_bug.cgi?id=286882
rdar://144045558

Reviewed by Alan Baradlay.

This patch aligns WebKit with Web Specification [1]:

[1] https://www.w3.org/TR/CSS2/visuren.html#dis-pos-flo

In this patch, we return computed value of `float: none` irrespective of
whether the display is `none` or not. Which is aligned with specification,
which mentions:

"If 'display' has the value 'none', then 'position' and 'float' do not apply."

and similarly if position is absolute or fixed (hasOutOfFlowPosition) function,
then it should also be float value should be `none`.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(ComputedStyleExtractor::valueForPropertyInStyle):
* LayoutTests/imported/w3c/web-platform-tests/css/CSS2/floats/computed-float-position-absolute-expected.txt:

Canonical link: https://commits.webkit.org/290061@main
